### DISCUSSION TITLE
关于测试用例的请教。

### DISCUSSION THREAD

**Discussion by derekwin** [alibaba/xquic#251] [Q&A] [answered]
测试环境，ubuntu22  mininet  20mbps 40ms链路
几个问题：
1. 目前测试用例 -s 方式传送较大body的时候，是这种断续的，这里难道是每次有大小限制，所以分几次去传了？
![image](https://user-images.githubusercontent.com/23076088/197739573-4c56f37c-0ffd-40f0-9425-eef9b815f715.png)
server:bbr，client:bbr

我人为的固定并压低发送窗口：
![image](https://user-images.githubusercontent.com/23076088/197741715-6f293d31-e2c3-43f8-9eaa-a14936888063.png)
貌似这貌似这个波动的下限在20mbps这个场景就是在1MBps左右

进一步压低：
![image](https://user-images.githubusercontent.com/23076088/197742259-9edc7c1a-d170-49df-a84d-ce443e7abfdf.png)

这里还有一个疑问就是所有的算法貌似在这个测试用例跑的时候都不是很稳定（不像tcp里面的）
比如tcp里面cubic是这个效果(忽略前面的流没起来的那段)
![image](https://user-images.githubusercontent.com/23076088/197747029-bf22bf3f-a1e4-4816-b43b-0bbd1a04c612.png)

tcp bbr
![image](https://user-images.githubusercontent.com/23076088/197749663-ea618e0d-1060-4d52-841e-79c9f75583ac.png)

2. cubic算法和bbr算法的差距有些感觉不太对；内核tcp对比这两种算法，起码在吞吐上来说差距是没这么大的。
![image](https://user-images.githubusercontent.com/23076088/197740702-45b4d988-0bd4-403a-80c5-a1a0ec6df835.png)
server:cubic，client:cubic

3. 测试用例的-r这个是怎么用的呢？我直接-r文件路径，报读错误了。这块的-r难道不是指向一般文件的？这里文档里面也没有使用说明。望告知。

感谢！

**Comment by derekwin** [comment]
第三个我知道我的问题了，我-r的文件超出MAX_BUF_SIZE了，换了个小点的文件可以了。

**Comment by yangfurong** [comment, marked as answer]
1. BBR的实现不是标准的BBR，由一个线性增窗的逻辑，可能是导致BBR吞吐波动的原因，后续版本我们会加上开关把它关掉；
2. Cubic的实现有一些问题，内部版本已经修复，后续会同步上来；
3. 内部版本还有一版类似TCP内核cubic的实现，后续也会同步上来；

**Reply by derekwin** [reply]
感谢回复。希望早日同步 hhhh。

**Reply by derekwin** [reply]
> 1. BBR的实现不是标准的BBR，由一个线性增窗的逻辑，可能是导致BBR吞吐波动的原因，后续版本我们会加上开关把它关掉；
> 2. Cubic的实现有一些问题，内部版本已经修复，后续会同步上来；
> 3. 内部版本还有一版类似TCP内核cubic的实现，后续也会同步上来；

不好意思，我又想起来一个问题，就是测试用例本地回环测试的时候，速度并不快，我在ubuntu物理机本地最高跑到10Mbps左右，虚拟机里面6-7Mbps水平。我用picoquic和quicly测试回环都是妙传，也是符合逻辑的。所以我感觉测试用例可能是不是也有问题。

**Comment by derekwin** [comment]
补充一个：我之前测的方式也有问题，传大文件时，客户端在运行的时候，要-T把http3关掉。
关闭前：
![image](https://user-images.githubusercontent.com/23076088/198172557-402d8df1-4623-451a-9d59-946742894abf.png)
关闭后：
![image](https://user-images.githubusercontent.com/23076088/198172260-76b4b9ae-2bc7-4760-af4e-cf9807c5be96.png)

  当然这只是其中一个因素，在yangfurong回复我之后，我也关掉H3对前面反馈的问题进行了重测，问题是依然存在的，所以证实了yang的回复。(当前时间节点221027)
